File: srv.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"time"

	mgo "gopkg.in/mgo.v2"
	"gopkg.in/mgo.v2/bson"
)

type access struct {
	Token string
}

type response struct {
	Message string
}

type user struct {
	Token        string
	UID          string
	ExpiresAt    time.Time `bson:"expires_at"`
	PasswordHash string    `bson:"password_hash"`
}

func main() {

	http.HandleFunc("/", greet)
	err := http.ListenAndServeTLS(":8001", "/etc/letsencrypt/live/supriya.tech/fullchain.pem", "/etc/letsencrypt/live/supriya.tech/privkey.pem", nil)
	if err != nil {
		log.Fatal("ListenAndServe: ", err)
	}
	fmt.Println("LISTEN AND SERVE OK")
}

// POST handler
func greet(w http.ResponseWriter, r *http.Request) {
	var a access
	var u user
	w.Header().Set("Content-type", "application/json")
	// Allow requests originating only from the flask app
	w.Header().Set("Access-Control-Allow-Origin", "https://supriya.tech")
	body, err := ioutil.ReadAll(r.Body)
	if err != nil {
		fmt.Println(body)
	}

	if json.Unmarshal(body, &a) != nil {
		fmt.Println("Unmarshall Error")
	}

	session, err := mgo.Dial("localhost:27017")
	if err != nil {
		fmt.Println("Couldnot connect to the DB")
	}
	c := session.DB("auth").C("users")
	session.SetMode(mgo.Monotonic, true)
	err = c.Find(bson.M{"token": a.Token}).One(&u)
	if err != nil {
		http.Error(w, "user not found", 404)
	}

	// Validate session using the token
	now := bson.Now()
	expired := u.ExpiresAt
	delta := expired.Sub(now)
	if delta < 0 {
		http.Error(w, "session expired", 401)
	}

	resp := response{
		Message: fmt.Sprintf("Gopher says, Hello! %v", u.UID),
	}

	b, err := json.Marshal(resp)
	if err != nil {
		fmt.Println("Marshalling Error")
	}
	w.Write(b)

	if err != nil {
		panic("Couldnot open a connection to the DB")
	}
	defer session.Close()
}

File: srv.py
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
import json, os
from pymongo import Connection
from datetime import datetime
connection = Connection("localhost:27017")
import ssl

class S(BaseHTTPRequestHandler):
    # Allow requests originating only from the flask app
    def _set_headers(self):
        self.send_header('Content-type', 'text/html')
        self.send_header('Access-Control-Allow-Origin', 'https://supriya.tech')
        self.end_headers()

    def do_GET(self):
        self._set_headers()
        self.wfile.write("<html><body><h1>I'm Alive!</h1></body></html>")

    def do_HEAD(self):
        self._set_headers()

    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)
        post_data = json.loads(post_data)
        print(post_data)
        greeting = self.get_greeting(post_data["token"])

    def get_greeting(self, token):
        db = connection.auth.users
        found_user = db.find_one({'token': token})
        if found_user == None:
            self.send_response(404)
            self._set_headers()
            self.wfile.write("Error! User not found. Please register")
        # Session validation
        if found_user['expires_at'] < datetime.utcnow():
            self.send_response(401)
            self._set_headers()
            self.wfile.write("Error! Unauthorized. Please login.")
        self.send_response(200)
        self._set_headers()
        resp = {}
        resp["Message"] = "Hello " + found_user["uid"] + "! Python is pleased to meet you!"
        self.wfile.write(json.dumps(resp))


def run(server_class=HTTPServer, handler_class=S, port=8000):
    server_address = ('', port)
    httpd = server_class(server_address, handler_class)
    httpd.socket = ssl.wrap_socket(httpd.socket, certfile='/etc/letsencrypt/live/supriya.tech/fullchain.pem', keyfile='/etc/letsencrypt/live/supriya.tech/privkey.pem', server_side=True, ssl_version=ssl.PROTOCOL_TLSv1_2)
    print 'Starting httpd...'
    httpd.serve_forever()

if __name__ == "__main__":
    from sys import argv

    if len(argv) == 2:
        run(port=int(argv[1]))
    else:
        run()
